Read :event_host setting at runtime

aaronrenner opened the pull request:
Currently, the `:event_host` setting is read at compile-time and can't be changed at runtime via settings in sys.config. This changes the code so the `:event_host` setting is read at runtime instead of compile time.

koudelka commented on lib/riemann/helpers/event.ex:
```diff
@@ -60,6 +56,14 @@ defmodule Riemann.Helpers.Event do
+      end
+
+      defp event_host_setting do
+        Application.get_env(:riemann, :event_host)
```
Ideally, i'd prefer not to call this every time we build an event (since i believe that `Application.get_env` is a `GenServer.call`).

If you can think of other way to cache it, I'd prefer that, but if not, it might not be a huge deal.

aaronrenner committed:
Adding test coverage for upcoming refactoring

aaronrenner committed:
Read :event_host setting at runtime

Currently, the :event_host setting is read at compile-time and can't be
changed at runtime via settings in sys.config. This changes the code so
the :event_host setting is read at runtime instead of compile time.

I also moved the logic for building an event into Riemann.Helpers.Event
instead of creating a bunch of private functions that pollute the module
that `uses` this module.

aaronrenner commented:
I reworked the code so it only tries to figure out the current hostname if one isn't provided. I also moved the `build/1` logic into `Riemann.Helpers.Event` so the module that's including this code isn't polluted with our private functions.

It should be good to merge, if it looks good to you.

koudelka commented on lib/riemann/helpers/event.ex:
```diff
@@ -59,7 +40,46 @@ defmodule Riemann.Helpers.Event do
+    end
+  end
+
+  defp set_metric_pb_fields(map) do
```
what do you think of splitting this up into a function match instead?

`defp set_metric_pb_fields(%{metric: i} = map) when is_integer(i) do` etc...

koudelka commented on lib/riemann/helpers/event.ex:
```diff
@@ -17,28 +17,9 @@ defmodule Riemann.Helpers.Event do
-        end
-        |> Map.to_list
-        |> new()
+        Riemann.Helpers.Event.build(dict, __MODULE__)
```
how about `unquote(__MODULE__).build(dict, __MODULE__)` instead?

aaronrenner replied:
That makes sense for maintainability. Tricky that `__MODULE__` means two different things on the same line. I'll make sure to add a comment.

koudelka commented on lib/riemann/helpers/event.ex:
```diff
@@ -59,7 +40,46 @@ defmodule Riemann.Helpers.Event do
+    event_host_setting() || machine_hostname()
+  end
+
+  defp set_attributes_field(map) do
```
same suggestion as for line 69?

aaronrenner replied:
The tricky part to pattern matching this is the `:attributes` field can either be `nil` or not set. If it's nil or not set, we return the map, if it's anything else send it to `Attribute.build/1`. If I wanted to use pattern matching without any other adjustments to the code, it might have to look something like this to keep the same behavior:
```elixir
defp set_attributes_field(%{attributes: nil}=map), do: map
defp set_attributes_field(%{attributes: a}=map) do
  Map.put(map, :attributes, Attribute.build(a))
end
defp set_attributes_field(map), do: map
```

 To make this prettier, I'm thinking I could update `build/2` to default that attribute to `nil`:
```elixir
def build(args, mod) do
  args
  |> Enum.into(%{})
  |> Map.put_new(:attributes, nil)
  |> # ...
end
```
Then I could pattern match like this:
```elixir
defp set_attributes_field(%{attributes: nil}=map), do: map
defp set_attributes_field(%{attributes: a}=map) do
  Map.put(map, :attributes, Attribute.build(a))
end
```

I took the second approach with the `set_metric_pb_fields` function on my next PR. What do you think?

koudelka replied:
Hm, how about instead of modifying `build/2`, just this?

```elixir
defp set_attributes_field(%{attributes: a} = map) when not is_nil(a) do
  Map.put(map, :attributes, Attribute.build(a))
end
# catches both nil and missing cases
defp set_attributes_field(map), do: map
```

aaronrenner replied:
That's a much better solution. I've been staring at this too long. 😄

koudelka replied:
Backseat driving is 20/20 :D

aaronrenner committed:
Cleanup case statements and add InvalidMetricError

This migrates to set_attribute_field and set_metric_pb_fields to use
multiple function definitions instead of case statements. This also
raises an InvalidMetricError when the metric is of an unexpected
data type.

aaronrenner commented:
I just cleaned up those case statements and added an `InvalidMetricError` when the metric is an invalid data type. `nil` still raises the old `ArgumentError`, but I'm going to address that in my next PR where I'll allow events with nil metrics.
